File: 2025/day_01/s.jl
function instr_to_amount(instr)
  amount = parse(Int, instr[2:end])
  if instr[1] == 'L'
    return -amount
  else
    return amount
  end
end

data = read("./input.txt", String) |> split .|> instr_to_amount

function part1()
  pos = 50
  count = 0
  for n in data
    pos = (pos+n) % 100
    count += pos==0
  end
  return count
end

function part2()
  pos = 50
  count = 0
  for n in data
    for _=1:abs(n)
      pos += sign(n)
      count += pos==0
      if pos < 0
        pos += 100
      end
      if pos == 100
        pos -= 100
        count += 1
      end
    end
    # println(pos, " - ", count)
  end
  return count
end

println(part1())
println(part2())

File: 2025/day_01/s.hs
parseInstruction :: String -> Int
parseInstruction (dir : amountStr)
  | dir == 'L' = -amount
  | dir == 'R' = amount
  where
    amount = read amountStr :: Int

countZeros :: (Int, Int) -> Int -> (Int, Int)
countZeros (pos, nZeros) move = (nextPos, nZeros + eqZero)
  where
    nextPos = (pos + move) `mod` 100
    eqZero = if nextPos == 0 then 1 else 0

part1 :: [Int] -> Int
part1 moves = count
  where
    start = 50
    (_finalPos, count) = foldl countZeros (start, 0) moves

countInterZeros :: (Int, Int) -> Int -> (Int, Int)
countInterZeros (pos, nZeros) move = (nextPos, nZeros + interZeros)
  where
    nextPos = (pos + move) `mod` 100
    interZeros = length (filter (== 0) (map (`mod` 100) (tail [pos, pos + (move `div` abs move) .. pos + move])))

part2 :: [Int] -> Int
part2 moves = count
  where
    start = 50
    (_finalPos, count) = foldl countInterZeros (start, 0) moves

main :: IO ()
main = do
  input <- lines <$> readFile "input.txt"
  let moves = map parseInstruction input
  let count1 = part1 moves
  print count1

  let count2 = part2 moves
  print count2
